windows/RNPrint/RNPrint.h

```diff
@@ -0,0 +1,402 @@
+              } else
+              {
+                printTask.Options().Orientation(PrintOrientation::Portrait);
+              }
+
+              printTask.Completed([=](PrintTask const& sender, PrintTaskCompletedEventArgs const& args)
+                {
+                  cleanUp();
+                  switch (args.Completion())
+                  {
+                  case PrintTaskCompletion::Failed:
+                    capturedPromise.Reject("Failed to print.");
+                    break;
+                  case PrintTaskCompletion::Abandoned:
+                    capturedPromise.Reject("Printing Abandoned");
+                    break;
+                  default:
+                    capturedPromise.Resolve(options.jobName);
+                    break;
+                  }
+                }
+              );
+            }
+          );
+
+          co_await PrintManager::ShowPrintUIAsync();
+        }
+      }
+    }
+
+    REACT_METHOD(Print, L"print")
+      void Print(JSValueObject&& options, ReactPromise<JSValue> promise) noexcept
+    {
+      bool ok_to_go_ahead = true;
+      {
+        // Only support one print request at a time.
+        std::lock_guard<std::mutex> guard(printSync);
+        if (servingPrintRequest)
+        {
+          ok_to_go_ahead = false;
+        } else
+        {
+          servingPrintRequest = true;
+        }
+      }
+      if (!ok_to_go_ahead)
+      {
+        promise.Reject("Another print request is already being served.");
+        return;
+      }
+
+      RNPrintOptions printOptions;
+
+      if (options.find("html") != options.end())
+        printOptions.html = options["html"].AsString();
+
+      std::optional<std::string> filePath = std::nullopt;
+      if (options.find("filePath") != options.end())
+        printOptions.filePath = options["filePath"].AsString();
+
+      printOptions.isLandscape = (options.find("isLandscape") != options.end() ? options["isLandscape"].AsBoolean() : false);
+      printOptions.jobName = (options.find("jobName") != options.end() ? options["jobName"].AsString() : defaultJobName);
+
+      xaml::FrameworkElement root{ nullptr };
+
+      auto window = xaml::Window::Current();
+      root = window.Content().as<xaml::FrameworkElement>();
+      printCanvas = searchForPrintCanvas(root);
+
+      if (!printCanvas)
+      {
+        promise.Reject("The XAML Canvas named \"RNPrintCanvas\" was not found.");
+        return;
+      }
+
+      auto asyncOp = PrintAsyncHelper(printOptions, promise);
+      asyncOp.Completed([=](auto action, auto status)
+        {
+          // Here we handle any unhandled exceptions thrown during the
+          // asynchronous call by rejecting the promise with the error code
+          if (status == winrt::Windows::Foundation::AsyncStatus::Error)
+          {
+            cleanUp();
+            std::stringstream errorCode;
+            errorCode << "0x" << std::hex << action.ErrorCode() << std::endl;
```

Comment:
undo the std::hex at the end?

Reply:
Do you mean the `std::endl`?

I've used this code as a reference:
https://github.com/microsoft/react-native-windows-samples/blob/c1fe1ad0395d905c790b97edde76155efa9cced9/samples/NativeModuleSample/cppwinrt/windows/NativeModuleSample/AsyncMethodExamples.h#L68

But now that I think about it, that `std::endl` might not make sense.

Reply:
I think it's a bug in the RNW code too :) basically std::hex changes state of the cout object, so any ints that come afterwards will be formatted as hex. AFAICT, endl only inserts a newline and doesn't reset the formatting options, so if someone prints a number to cout before anyone does << std::hex, it will show up in decimal, but if someone does it after this code runs, it will print as hex. Therefore it is unpredictable and we should reset to the default at the end of the line (<< std::dec << std::endl instead of std::endl)

Reply:
Thank you. SGTM :)